File: Strivers-A2Z-DSA-Sheet/05.Recursion/3.Try Out All Combos/python/06.Word Break.py
"""
=============================================================================
  QUESTION: 139. Word Break (LeetCode)
=============================================================================

  Can string s be segmented into dictionary words?
  Example: s="leetcode", dict=["leet","code"] → True

  APPROACH: Backtracking — same pattern as Palindrome Partitioning.
  Try every substring from current index. If in dict → recurse on remaining.

  DRY RUN "leetcode":
    "l"→no "le"→no "lee"→no "leet"→YES → solve(4)
      "c"→no "co"→no "cod"→no "code"→YES → solve(8)
        index >= len → return True ✓

  DRY RUN "catsandog":
    "cat"→YES → "sandog": "sand"→YES → "og": no match → False
    "cats"→YES → "andog": "and"→YES → "og": no match → False ✓

  Time: O(2^N) brute, O(N²) with memo. Space: O(N)
=============================================================================
"""


def solve(index, s, word_set):
    if index >= len(s):
        return True

    for i in range(index, len(s)):
        word = s[index:i + 1]
        if word in word_set and solve(i + 1, s, word_set):
            return True
    return False


def word_break(s, word_dict):
    word_set = set(word_dict)
    return solve(0, s, word_set)


# Driver Code
if __name__ == "__main__":
    print(word_break("leetcode", ["leet", "code"]))        # True
    print(word_break("applepenapple", ["apple", "pen"]))    # True
    print(word_break("catsandog", ["cats","dog","sand","and","cat"]))  # False

File: Strivers-A2Z-DSA-Sheet/05.Recursion/3.Try Out All Combos/javascript/06.Word Break.js
/*
=============================================================================
  QUESTION: 139. Word Break (LeetCode)
=============================================================================

  Given string s and dictionary, return true if s can be segmented into
  space-separated dictionary words.

  Examples:
    "leetcode", ["leet","code"] → true
    "catsandog", ["cats","dog","sand","and","cat"] → false

=============================================================================
  APPROACH: Backtracking — Try All Word Splits — O(2^N) Time, O(N) Space
=============================================================================

  Same pattern as Palindrome Partitioning:
  At each index, try every substring s[index..i].
  If it's in dict → recurse on remaining. If remaining resolves → true.

  Dry Run "leetcode", dict={"leet","code"}:
    solve(0) → "l" ✗ | "le" ✗ | "lee" ✗ | "leet" ✓
      → solve(4) → "c" ✗ | "co" ✗ | "cod" ✗ | "code" ✓
        → solve(8) → index >= length → true ✓

  NOTE: Add memoization for O(N²) optimization.

=============================================================================
*/

function wordBreak(s, wordDict) {
    let dict = new Set(wordDict);

    function solve(index) {
        if (index >= s.length) return true;

        for (let i = index; i < s.length; i++) {
            let word = s.substring(index, i + 1);
            if (dict.has(word) && solve(i + 1)) return true;
        }
        return false;
    }

    return solve(0);
}

// ==========================================================================
// DRIVER CODE
// ==========================================================================
console.log(wordBreak("leetcode", ["leet", "code"]));                       // true
console.log(wordBreak("applepenapple", ["apple", "pen"]));                   // true
console.log(wordBreak("catsandog", ["cats","dog","sand","and","cat"]));       // false
